### ISSUE TITLE
Some questions of multi-head self-attention relation distillation

### ISSUE THREAD

**Issue by RyanHuangNLP** [microsoft/unilm#354]
**Describe**
Model I am using (MiniLMv2):
![image](https://user-images.githubusercontent.com/49582480/123972016-16ec6600-d9ed-11eb-8b21-163fe9777b77.png)
@WenhuiWang0824 @donglixp Thanks for updating your paper with a more intuitive schematic, I still have some implementation details I would like to ask for, please allow me to describe the problem (mark with **Q**) in the following code:

```
# If the teacher query has 16 head size, and batch size is 16, seq length is 64, head per size is 64
q_teacher = torch.rand((16, 16, 64, 64))
# Q1: I found that the split operation both change the origin head size and head per size, Isn't it a bit confusing to call it split?
q_teacher = torch.reshape(q_teacher, (16, -1)) 
q_teacher = torch.reshape(q_teacher, (16, 64, 64, -1)) # now the head per size is 16

# If the student query has 12 head size, and batch size is 16, seq length is 64, head per size is 64
q_student = torch.rand((16, 12, 64, 64))
q_student = torch.reshape(q_student, (16, -1))
q_student = torch.reshape(q_student, (16, 64, 64, -1)) # now the head per size is 12

qqt_relation_scores = torch.matmul(q_teacher, q_teacher.transpose(-1, -2))
qqt_relation_scores = qqt_relation_scores / math.sqrt(16) # now the head per size is 16
qqt_relation_scores = qqt_relation_scores + attention_mask # Q2: here relation scores need to plus attention mask?
# Q3: Use relation probrability with or without dropout operation
qqt_relation_probs = nn.Softmax(dim=-1)(qqt_relation_scores)
qqt_relation_probs = nn.dropout(qqt_relation_probs)

qqs_relation_scores = torch.matmul(q_student, q_student.transpose(-1, -2))
qqs_relation_scores = qqs_relation_scores / math.sqrt(12) # now the head per size is 12
qqs_relation_scores = qqs_relation_scores + attention_mask 
qqs_relation_probs = nn.Softmax(dim=-1)(qqt_relation_scores)
qqs_relation_probs = nn.dropout(qqs_relation_probs)

# last compute the kl loss between qqt_relation_probs and qqs_relation_probs
```

**Comment by WenhuiWang0824** [comment]
Hi @RyanHuangNLP 

For Q1: I think you need to concatenate the vector of different heads first and then split the concatenated vector (16, 64 (seq_len), 1024) according to the desired size (16, 64, 64, -1).
For Q2: Yes
For Q3: I do not add dropout here.

Thanks

**Comment by RyanHuangNLP** [comment]
@WenhuiWang0824 Thanks! For Q1: It should be concat and split after the hidden state -> linear operation

**Comment by LiuChiachi** [comment]
Hi~  @WenhuiWang0824  @donglixp Thanks for your contribution! After reading the MINILMv2 paper I think the "concatenation and split" operation is just performed at loss function, and it wouldn't change the multihead attention itself in student or teacher network structure. Is that right? Why @RyanHuangNLP 's sample code seems like the forward network has been changed?

Thanks.
